### PR TITLE
ZBar: add package

### PR DESCRIPTION
**Pull request by Doekin**
### Description

ZBar is an open-source software suite for reading barcodes from various sources, including webcams. Its original development stopped in 2012, and [mchehab](https://github.com/mchehab/zbar) took the task of keeping it updated with the V4L2 API.

### Configuration

ZBar uses autotools and includes various components such as Python bindings, video input, and GUI widgets. For simplicity and portability, it has been ported to xmake, however, only essential configuration options are translated.

### CODE REVIEW

File: packages/z/zbar/port/config.h [hunk 1]
```diff
@@ -0,0 +1,480 @@
...
+
+/* Define to 1 if you have the <malloc.h> header file. */
+/* #undef HAVE_MALLOC_H */
+
+/* Define to 1 if you have the 'memchr' function. */
+#define HAVE_MEMCHR 1
+
+/* Define to 1 if you have the 'memmove' function. */
+#define HAVE_MEMMOVE 1
+
```
Comment (waruqi) [inline packages/z/zbar/port/config.h:166]: we should use add_configfiles + configvar

https://github.com/xmake-io/xmake-repo/blob/5ca5ddc2eb11f8efe35a683a5e4dc701a707d4b3/packages/l/libiconv/port/xmake.lua#L229

Comment (Doekin) [reply packages/z/zbar/port/config.h:166]: I will try `io.gsub("config.h.in", "# ?undef (.-)\n", "${define %1}\n")`

File: packages/z/zbar/port/config.h [hunk 2]
```diff
@@ -0,0 +1,480 @@
...
+/* whether to build support for SQ Code */
+#define ENABLE_SQCODE 1
+
+/* Define to 1 if you have the 'alarm' function. */
+#define HAVE_ALARM 1
+
+/* Define to 1 if you have 'alloca', as a function or macro. */
+/* #undef HAVE_ALLOCA */
+
```
Comment (waruqi) [inline packages/z/zbar/port/config.h:46]: if you use config.h.in, you can remove config.h

File: packages/z/zbar/xmake.lua [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    else
+        add_deps("libiconv")
+    end
+
+    on_install("!iphoneos and !windows", function (package)
+        os.cp(path.join(package:scriptdir(), "port", "config.h.in"), "include/config.h.in")
+        io.gsub("include/config.h.in", "# ?undef (.-)\n", "${define %1}\n")
+        os.cp(path.join(package:scriptdir(), "port", "xmake.lua"), "xmake.lua")
+        
+        local config = { vers = package:version_str() }
```
Comment (waruqi) [inline packages/z/zbar/xmake.lua:33]: rename to configs

File: packages/z/zbar/xmake.lua [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+        -- get LIB_VERSION from configure.ac
+        -- format: AC_SUBST([LIB_VERSION], [3:0:3])
+        configure_ac = io.readfile("configure.ac")
```
Comment (waruqi) [inline packages/z/zbar/xmake.lua:37]: local configure_ac

File: packages/z/zbar/xmake.lua [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+        for _, key in ipairs({"LIB_VERSION"}) do
+            local value = configure_ac:match("AC_SUBST%(%[" .. key .. "%],%s*%[(.-)%]%)")
```
Comment (waruqi) [inline packages/z/zbar/xmake.lua:39]: add spaces ` , `

File: packages/z/zbar/xmake.lua [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+    elseif is_plat("macosx") then
+        add_extsources("brew::zbar")
+    end
+
+    if is_plat("linux", "bsd") then
+        add_syslinks("pthread")
+    end
+
+    if is_plat("macosx") then
+        add_deps("libiconv", {system = true})
```
Comment (waruqi) [inline packages/z/zbar/xmake.lua:23]: libiconv will find and use system libs first. so you need not add `system = true`

File: packages/z/zbar/port/xmake.lua [hunk 7]
```diff
@@ -0,0 +1,189 @@
+add_rules("mode.debug", "mode.release")
+
+add_requires("libiconv")
+
+-- add options
+option("enable_codebar")
```
Comment (waruqi) [inline packages/z/zbar/port/xmake.lua:6]: remove enable_ prefix

File: packages/z/zbar/port/xmake.lua [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+                {name = "sqcode",  description = "SQ name"},
+                {name = "pdf417",  description = "PDF417 symbology (incomplete)"}}
+for _, code in ipairs(codes) do
+    option(code.name)
+        -- set_default(false)
+        add_deps("enable-codes")
+        set_description("whether to build support for " .. code.description)
+
+        on_check(function (option)
+            local enabled_codes = option:dep("enable-codes"):value()
```
Comment (waruqi) [inline packages/z/zbar/port/xmake.lua:29]: I didn't see it in tools.xmake.install.

File: packages/z/zbar/port/xmake.lua [hunk 9]
```diff
@@ -0,0 +1,164 @@
+add_rules("mode.debug", "mode.release")
+
+add_requires("libiconv")
+
+-- add options
+option("enable-codes")
```
Comment (waruqi) [inline packages/z/zbar/port/xmake.lua:6]: please  remove `enable-` prefix

File: packages/z/zbar/xmake.lua [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+    elseif is_plat("linux") then
+        add_extsources("pacman::zbar", "apt::libzbar-dev")
+    elseif is_plat("macosx") then
+        add_extsources("brew::zbar")
+    end
+
+    if is_plat("linux", "bsd") then
+        add_syslinks("pthread")
+    end
+
```
Comment (waruqi) [inline packages/z/zbar/xmake.lua:21]: no any add_configs to set `xxx codes` options in port/xmake.lua

Comment (Doekin) [reply packages/z/zbar/xmake.lua:21]: Oh, we do need configs in the package description.
By the way, which form is better?
1. `xrepo install -f "qr=true" zbar`
2. `xrepo install -f "codes=qr" zbar`

For reference, the original `configure` shell script uses options like the following:
`./configure --disable-video --without-gtk --without-python --enable-shared=yes --enable-static=yes --with-pic --enable-codes=ean,databar`

Comment (waruqi) [reply packages/z/zbar/xmake.lua:36]: you can use single codes option/config

```lua
add_requires("xxx", configs = {codecs = {"qr", "ean", ...})
``

File: packages/z/zbar/port/xmake.lua [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+    set_configvar("PACKAGE_VERSION", get_config("vers"))
+    set_configvar("PACKAGE_STRING", "zbar " .. get_config("vers"))
+    
+    local vers = get_config("vers"):split("%.")
+    major_ver = vers[1] or ""
+    minor_ver = vers[2] or ""
+    patch_ver = vers[3] or ""
+    set_configvar("ZBAR_VERSION_MAJOR", major_ver, {quote = false})
+    set_configvar("ZBAR_VERSION_MINOR", minor_ver, {quote = false})
+    set_configvar("ZBAR_VERSION_PATCH", patch_ver, {quote = false})
```
Comment (waruqi) [inline packages/z/zbar/port/xmake.lua:26]: use set_versions, xmake has builtin vars in config.h.in

see https://github.com/tboox/tbox/blob/1c3c01b1757febc4ade5ae16ce6a90c624b07d3c/src/tbox/tbox.config.h.in#L5-L9

File: packages/z/zbar/xmake.lua [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+    if is_plat("mingw") and is_subhost("msys") then
+        add_extsources("pacman::zbar")
+    elseif is_plat("linux") then
+        add_extsources("pacman::zbar", "apt::libzbar-dev")
+    elseif is_plat("macosx") then
+        add_extsources("brew::zbar")
+    end
+
+    add_configs("symbologies", {description = "Select symbologies to compile", default = {"ean","databar","code128","code93","code39","codabar","i25","qrcode","sqcode"}, type = "table"})
```
Comment (waruqi) [inline packages/z/zbar/xmake.lua:18]: add spaces ` , `

File: packages/z/zbar/port/xmake.lua [hunk 13]
```diff
@@ -0,0 +1,120 @@
+add_rules("mode.debug", "mode.release")
+
+add_requires("libiconv")
+
+-- add options
+option("symbologies")
+    -- set_default({"ean","databar","code128","code93","code39","codabar","i25","qrcode","sqcode"})
```
Comment (waruqi) [inline packages/z/zbar/port/xmake.lua:7]: remove comment

File: packages/z/zbar/port/xmake.lua (hunk 11 above)
Comment (Doekin) [reply packages/z/zbar/port/xmake.lua:27]: The `config.h.in` file contains content like the following:
```
/* Program major version (before the '.') as a number */
#undef ZBAR_VERSION_MAJOR

/* Program minor version (after '.') as a number */
#undef ZBAR_VERSION_MINOR

/* Program patch version (after the second '.') as a number */
#undef ZBAR_VERSION_PATCH
```
I'm not sure how to handle this.

Comment (waruqi) [reply packages/z/zbar/port/xmake.lua:27]: see https://github.com/tboox/tbox/blob/1c3c01b1757febc4ade5ae16ce6a90c624b07d3c/src/tbox/tbox.config.h.in#L5-L9

Comment (Doekin) [reply packages/z/zbar/port/xmake.lua:27]: Should I edit `config.h.in` to the following?
```
/* Program major version (before the '.') as a number */
#define ZBAR_VERSION_MAJOR ${VERSION_MAJOR}

/* Program minor version (after '.') as a number */
#define ZBAR_VERSION_MINOR ${VERSION_MINOR}

/* Program patch version (after the second '.') as a number */
#define ZBAR_VERSION_PATCH ${VERSION_ALTER}
```

Comment (waruqi) [reply packages/z/zbar/port/xmake.lua:27]: right

File: packages/z/zbar/port/xmake.lua [hunk 14]
```diff
@@ -0,0 +1,112 @@
...
+
+option("vers")
+    set_default("")
+    set_showmenu(true)
+option_end()
+if has_config("vers") then
+    set_version(get_config("vers"))
+
+    set_configvar("PACKAGE_VERSION", get_config("vers"))
+    set_configvar("PACKAGE_STRING", "zbar " .. get_config("vers"))
```
Comment (waruqi) [inline packages/z/zbar/port/xmake.lua:18]: and here

**Pull request merged**